### PR TITLE
Add MoNe experts pruning support

### PR DESCRIPTION
**Pull request by ieBoytsov**
## Purpose

Mixture of Novices and Experts (MoNe) is an expert pruning method accepted at ICLR-2026 https://arxiv.org/pdf/2507.00390
 
The method computes statistics (access frequency and output variance) about experts during calibration, identifies redundant experts and replaces them with constant vectors which are their unbiased estimators. MoNe is planned to be supported in the llm-compressor https://github.com/vllm-project/llm-compressor/pull/2906 

Implementation
- Defines a generic expert-substitution format in compression_config.
- Supports constant-v1, where a substituted expert returns a stored constant vector while keeping its original router weight.
- Automatically detects compatible MoE layers and loads only retained expert weights, preserving the GPU memory saving.
- Provides a default path for compatible decomposed MoE backends by sending substituted routes with zero weights.
- Validates checkpoint metadata and tensor loading, including tensor-parallel correctness.
Models using the standard FusedMoEFactory and RoutedExperts loading path do not need architecture-specific changes.

### CODE REVIEW

**Comment by ieBoytsov** [comment]
An example of benchmarking minimax-m2.5 where 25% pruned model yields better metrics on workloads where KV-cache pressure is high. MoNe has more KV-cache space given experts consume less GPU memory now. This translates to perf improvements.


UPD: outdated as the scope of this PR is limited to deepseek/GLM. Minimax to be added in follow up PRs.

|comparison_metrics

**Comment by ieBoytsov** [comment]
FYI @mgoin @HDCharles

**Comment by ieBoytsov** [comment]
I also tested GLM-5.2-FP16 vs GLM-5.2-FP16-MoNe (25% pruned experts), model quality on GSM8K was 95.5% original vs 93.3% MoNe using just 100 calibration samples and can be further improved with more samples.

From the deployment perspective GLM-5.2 can be used on 8x B200 only with CPU offload while MoNe model can sit there with no CPU offload and is much faster.

Also MoNe is orthogonal to different model optimisations techniques and can be further combined with quantization or/and spec dec.

**Comment by milakov** [comment]
The design document describes this as an “initial implementation” and constant-v1 as the “first supported executor,” but it does not explicitly explain that this is intended as the first self-contained PR in a sequence of incremental MoNE improvements. Could we add a short scope/roadmap note, for example:

“This PR is the first self-contained step toward broader MoNE support. It introduces an algorithm-independent expert-substitution checkpoint contract and implements the minimal constant-v1 inference path for unquantized DeepSeek/GLM models using the Triton backend and tensor parallelism. It intentionally does not support additional model families, replacement formats, quantized experts, expert/data parallelism, or other MoE backends. Those capabilities can be added incrementally in follow-up PRs without changing the version-1 checkpoint contract. MoNE checkpoint production and calibration will be implemented separately in llm-compressor.”

This would clarify the planned sequencing while emphasizing that the present PR is independently useful and does not depend on those follow-ups.

**Comment by milakov** [comment]
Consider adding a true configuration-to-model-loading test. The existing tests mostly construct the parser and `FusedMoE` objects directly. A test should load a minimal checkpoint/config through `ModelConfig` and the normal weight loader. This is especially important because current vLLM treats every `compression_config` as possible quantization metadata and inserts an empty `quant_method`; `_verify_quantization` currently turns that into `None`, so it probably works, but only through incidental behavior in the quantization path. That interaction should be explicitly handled or at least protected by an end-to-end regression test. See the current [compression normalization](https://github.com/vllm-project/vllm/blob/main/vllm/transformers_utils/model_arch_config_convertor.py) and [quantization verification](https://github.com/vllm-project/vllm/blob/main/vllm/config/model.py)

**Comment by milakov** [comment]
Replace runtime `assert` statements used for configured module-path validation with descriptive `ValueError`s. Assertions can disappear under optimized Python and give poor checkpoint diagnostics.

**Comment by milakov** [comment]
Consider normalizing terminology: the public schema says “substitution,” while several runtime names still say `expert_replacement`. This is not a correctness problem, but this first PR is where the API vocabulary becomes established.

**Comment by milakov** [comment]
Add at least a small performance measurement. The implementation introduces a constant-output path, extra buffer, and addition for affected layers. A benchmark would establish the initial cost even if optimization is deferred.

**Comment by ieBoytsov** [comment]
@milakov thanks for the review. I added changes to address your comments (changes are incorporated into existing commit):

* Added explicit handling for transform-only compression_config metadata so it is not treated as quantization, while preserving configurations that combine transforms and quantization. Added an end-to-end regression that loads a minimal DeepSeek checkpoint through ModelConfig and DefaultModelLoader.

* Replaced runtime assertions used for configured module-path validation with descriptive ValueErrors and added tests for invalid and mismatched paths.

* Normalized the runtime, tests, filenames, and documentation around the “expert substitution” terminology.

* Rebased the branch onto the latest upstream/main and resolved the conflicts.

I also tested inference with the official deepseek-v2-lite pruned [model](https://huggingface.co/MoNE-Pruning/DeepSeek-V2-Lite-MoNE-48-c4-1000) from MoNe paper authors and was able to reproduce their eval results (I used PIQA dataset for quick check and the score was almost identical). 

I will add performance benchmarks soon.

**Comment by ieBoytsov** [comment]
Perf benchmarking. FLOPS improvements from pruning described on this plot are expected to be moderate (while most of efficiency comes from GPU memory reduction and ability to deploy models using less amount of GPUs)


Model | Checkpoint size | Saved vs original | KV-cache capacity
-- | -- | -- | --
Original | 29.26 GiB | — | 4.411M tokens
25% pruned | 22.55 GiB | 6.71 GiB (22.9%) | 4.639M tokens (+5.2%)
50% pruned | 15.85 GiB | 13.41 GiB (45.8%) | 4.870M tokens (+10.4%)

Tested under `--gpu-memory-utilization 0.9`


comparison_metrics_robust

**Comment by vadiklyutiy** [comment]
Two things are bothering me a little: 
1. We have to add manually support for every model
2. We have to add manually support every MoE backend

This really limits adoption of MoNE. 

I am wondering could we design implementation to eliminate 2 above concerns?

For example instead of support in every backend, we may filter tokens with dummy weights early and send only tokens with real weights.

**Comment by ieBoytsov** [comment]
> Two things are bothering me a little:
> 
> 1. We have to add manually support for every model
> 2. We have to add manually support every MoE backend
> 
> This really limits adoption of MoNE.
> 
> I am wondering could we design implementation to eliminate 2 above concerns?
> 
> For example instead of support in every backend, we may filter tokens with dummy weights early and send only tokens with real weights.

Hi @vadiklyutiy thank you! 

I would say it is possible to support stadardised MoE models and most of backends using a non-optimised execution. This still gives a memory win. On top of that it is possible to further optimise specific backeds to make them faster.

Here is the proposal:


For models, `FusedMoEFactory` would automatically discover substitution metadata using the module prefix it already receives. Generic loading and validation hooks would handle the substitution tensors. This should make the feature work by default for most models using the standard fused-MoE path, without modifying each model implementation.

For backends, a generic adapter would run after routing. It would compute the constant contribution, map retained logical experts to compact physical IDs, and replace substituted routes with valid IDs carrying zero weights. Most decomposed MoE backends could then execute the result without knowing anything about expert substitution.

This default path would be correct but might still execute unnecessary GEMMs for the zero-weight routes. The current Triton implementation would demonstrate an optimized path: it recognizes and removes those routes before scheduling GEMMs. Other backends could add the same optimization later, but it would not be required for correctness or initial model support.

Would a generic default path for most models and decomposed backends, with Triton as the first optimized implementation, address your concerns? 

FYI @milakov

**Comment by milakov** [comment]
> For backends, a generic adapter would run after routing. It would compute the constant contribution, map retained logical experts to compact physical IDs, and replace substituted routes with valid IDs carrying zero weights. Most decomposed MoE backends could then execute the result without knowing anything about expert substitution.

This one looks good to me. After all, MoNE is about replacing rarely used experts with novices, so running dummy GEMMs for those experts by default shouldn't be a big hit to perf, (this is compared to optimized expert replacement aware implementations! There would be no perf penalty compared to today runtime of models anyway). MoNE reduces DRAM utilization, and this will work with default implementation. @vadiklyutiy , what do you think?

> For models, `FusedMoEFactory` would automatically discover substitution metadata using the module prefix it already receives. Generic loading and validation hooks would handle the substitution tensors. This should make the feature work by default for most models using the standard fused-MoE path, without modifying each model implementation.

@ieBoytsov, here, I think, it makes sense to specify clearly what are the conditions for the models to work as is with substituted weights? This might help us understand if PR could be improved here.

**Comment by vadiklyutiy** [comment]
> replace substituted routes with valid IDs carrying zero weights

Why do we need zero weights? Can we just forgot about substituted expert?

**Comment by ieBoytsov** [comment]
> > replace substituted routes with valid IDs carrying zero weights
> 
> Why do we need zero weights? Can we just forgot about substituted expert?

@vadiklyutiy  I think because backends expect a deterministic [num_tokens, top_k] inputs. In the MoNe case some of topK can be real experts and some are constant experts which need to be zeroed-out for the backend.

**Comment by ieBoytsov** [comment]
> @ieBoytsov, here, I think, it makes sense to specify clearly what are the conditions for the models to work as is with substituted weights? This might help us understand if PR could be improved here.

@milakov I think a model should work without model-specific substitution code when:
- It constructs its routed experts through `FusedMoEFactory` with a module prefix that can be matched to a configured substitution target.
- Its router produces standard logical top-k IDs and weights with router weights applied to expert outputs.
- The substitution value is in the same output space and uses the same scaling semantics as the routed experts.

**Comment by mgoin** [comment]
I agree we should optimize highly for simplicity of maintenance and compatibility with *all* moe kernels. We should be able to treat the constant experts kind of like shared experts and run them in an auxiliary cuda stream

File: vllm/model_executor/layers/fused_moe/expert_substitution.py [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+        safe_ids = torch.where(valid, topk_ids, torch.zeros_like(topk_ids)).long()
+        substitution_rows = self.substitution_index[safe_ids].long()
+        substitution_mask = valid & (substitution_rows >= 0)
+        # Keep router weights and constant accumulation in FP32 until the final cast.
+        gathered = self.values[substitution_rows.clamp_min(0)].float()
+        scales = topk_weights.float() * substitution_mask
+        result = torch.sum(gathered * scales.unsqueeze(-1), dim=1)
+        substitution_output[..., : result.shape[-1]].copy_(
+            result.to(substitution_output.dtype)
+        )
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/expert_substitution.py:325]: These are some large allocations: `gathered = [tokens, top_k, hidden_size]` upconverted to float32, and then we make another multiplied result. I would expect this activation overhead to be measurable and since the fused moe is wrapped in a custom op, torch.compile will not interact with this. We should either manually wrap or write a simple triton kernel to fuse the intermediate states.

File: vllm/model_executor/model_loader/base_loader.py [hunk 2]
```diff
@@ -52,6 +59,18 @@ def create_model(
...
+        self, model: nn.Module, weights: Iterable[tuple[str, torch.Tensor]]
+    ) -> set[str] | None:
+        """Load checkpoint tensors, including model-independent replacements."""
+        weights, substitution_params = intercept_expert_substitution_weights(
+            model, weights
+        )
+        loaded_weights = model.load_weights(weights)
+        if loaded_weights is not None:
+            loaded_weights.update(substitution_params)
+        return loaded_weights
```
Comment (mgoin) [inline vllm/model_executor/model_loader/base_loader.py:72]: This is the kind of thing we just can't easily do. Many places call `model.load_weights()` and it is understandably a pretty core external interface. For instance `GPUModelRunner.reload_weights()` call this directly and would break. The loading behavior needs to sit at a boundary shared by initial loading and reload, not individual loader paths.

File: vllm/model_executor/layers/fused_moe/expert_substitution.py [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+    duplicate = sorted(path for path, count in matched.items() if count > 1)
+    if missing or duplicate:
+        details = []
+        if missing:
+            details.append(f"unmatched targets: {missing}")
+        if duplicate:
+            details.append(f"targets matched more than once: {duplicate}")
+        raise ValueError(
+            "invalid expert_substitution model binding; " + "; ".join(details)
+        )
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/expert_substitution.py:422]: This won't work with pipeline parallelism since each process only owns some layers.

File: vllm/model_executor/model_loader/base_loader.py [hunk 4]
```diff
@@ -9,6 +10,12 @@
 from vllm.config import ModelConfig, VllmConfig
 from vllm.config.load import LoadConfig
 from vllm.logger import init_logger
+from vllm.model_executor.layers.fused_moe.expert_substitution import (
+    clear_expert_substitution_load_state,
+    intercept_expert_substitution_weights,
+    validate_expert_substitution_model,
+    validate_expert_substitution_weights_loaded,
+)
```
Comment (mgoin) [inline vllm/model_executor/model_loader/base_loader.py:18]: This simply shouldn't belong here. The universal loader shouldn't have to worry about this particular experimental feature, it is just the wrong layer to insert at. There are existing `runner_cls` or `routed_experts_cls` injection points that I think are worth considering before coming into common classes.

File: vllm/model_executor/layers/fused_moe/expert_substitution.py (hunk 1 above)
Comment (ieBoytsov) [reply vllm/model_executor/layers/fused_moe/expert_substitution.py]: @mgoin Wrapped it with decorator. On B200 with [tokens=1024, top_k=6, hidden=2048], peak additional allocation dropped from 108.13 MiB to 4.20 MiB.

File: vllm/model_executor/model_loader/base_loader.py (hunk 2 above)
Comment (ieBoytsov) [reply vllm/model_executor/model_loader/base_loader.py]: @mgoin Moved substitution-tensor handling into a model adapter’s load_weights() method covering both initial loading and direct reloads.

File: vllm/model_executor/layers/fused_moe/expert_substitution.py (hunk 3 above)
Comment (ieBoytsov) [reply vllm/model_executor/layers/fused_moe/expert_substitution.py]: @mgoin Fixed. Tested that DeepSeek-MoNE with PP=2 produces same result as with one GPU.

File: vllm/model_executor/model_loader/base_loader.py (hunk 4 above)
Comment (ieBoytsov) [reply vllm/model_executor/model_loader/base_loader.py]: @mgoin  Fixed. Used routed_experts_cls.
